// charity-frontend/src/App.tsx
import { useState, useMemo } from 'react';
import { 
  ConnectButton, 
  useSignAndExecuteTransaction, 
  useCurrentAccount,
  useSuiClientQuery 
} from '@mysten/dapp-kit';
import { Transaction } from '@mysten/sui/transactions';
import './App.css';

// CẤU HÌNH QUAN TRỌNG
const PACKAGE_ID = "0xbbbc797617b2441ba716ba333c45679587cd1b41e7ceaa32e73dddefedaf2954";
const MODULE_NAME = "charity_platform"; 
// LƯU Ý: Bạn cần thay ID này bằng ID CharityPlatform nhận được khi deploy
const PLATFORM_ID = "0xdd7e02bdc86f3acc2c9777d37835cfe4c597d030a526993138815f6cdfd28a96";
const SUI_CLOCK = "0x6"; 

function App() {
  const [searchTerm, setSearchTerm] = useState("");
  const [txHistory, setTxHistory] = useState<any[]>([]); 
  const [showModal, setShowModal] = useState(false);
  
  // UI vẫn giữ 4 ô nhập để thân thiện với người dùng
  const [formData, setFormData] = useState({ 
    title: '', 
    goal: '', 
    desc: '', 
    days: '7' // Số ngày gây quỹ
  });

  const account = useCurrentAccount();
  const { mutate: signAndExecute } = useSignAndExecuteTransaction();

  // --- LẤY DỮ LIỆU TỪ BLOCKCHAIN ---
  const { data: suiData, refetch } = useSuiClientQuery(
    'getOwnedObjects',
    {
      owner: account?.address || "",
      options: { showContent: true, showType: true }
    },
    { enabled: !!account }
  );

  const campaigns = useMemo(() => {
    if (!suiData?.data) return [];
    return suiData.data
      .map((obj: any) => {
        const content = obj.data?.content;
        if (content?.type !== `${PACKAGE_ID}::charity_platform::Campaign`) return null;

        const fields = content.fields;
        return {
          id: obj.data.objectId,
          title: fields.title,
          description: fields.description,
          goal: Number(fields.goal_amount) / 1_000_000_000,
          raised: Number(fields.current_amount) / 1_000_000_000,
          image: "https://via.placeholder.com/400x200",
          isActive: fields.is_active
        };
      })
      .filter((c): c is any => c !== null);
  }, [suiData]);

  // --- HÀM TẠO CHIẾN DỊCH (GỬI ĐỦ 6 THAM SỐ) ---
  const handleCreateCampaign = () => {
    if (!account) return alert("Vui lòng kết nối ví!");
    if (PLATFORM_ID.includes("CHÈN_ID")) return alert("Bạn chưa điền PLATFORM_ID vào code!");

    const txb = new Transaction();
    
    // Chuyển đổi dữ liệu cho đúng kiểu Move [cite: 12, 13]
    const goalInMist = BigInt(Math.floor(Number(formData.goal) * 1_000_000_000));
    const deadlineInSeconds = BigInt(Math.floor(Date.now() / 1000) + (Number(formData.days) * 86400));

    txb.moveCall({
      target: `${PACKAGE_ID}::${MODULE_NAME}::create_campaign`,
      arguments: [
        txb.object(PLATFORM_ID),         // 1. platform 
        txb.pure.string(formData.title),   // 2. title 
        txb.pure.string(formData.desc),    // 3. description 
        txb.pure.u64(goalInMist),          // 4. goal_amount 
        txb.pure.u64(deadlineInSeconds),   // 5. deadline 
        txb.object(SUI_CLOCK),             // 6. clock 
      ],
    });

    signAndExecute(
      { transaction: txb as any },
      {
        onSuccess: (result: any) => {
          alert("Tạo chiến dịch thành công!");
          setTimeout(() => refetch(), 3000);
          setShowModal(false);
          setTxHistory(prev => [{ digest: result.digest, type: "Tạo mới", time: new Date().toLocaleTimeString() }, ...prev]);
        },
        onError: (err: any) => alert("Lỗi VM: " + err.message)
      }
    );
  };

  // --- HÀM QUYÊN GÓP ---
  const handleDonate = (campaignId: string) => {
    if (!account) return alert("Vui lòng kết nối ví!");

    const txb = new Transaction();
    const [coin] = txb.splitCoins(txb.gas, [txb.pure.u64(1_000_000_000)]); // 1 SUI

    txb.moveCall({
      target: `${PACKAGE_ID}::${MODULE_NAME}::donate`,
      arguments: [
        txb.object(campaignId), // 1. campaign [cite: 18]
        coin,                   // 2. payment [cite: 18]
        txb.object(SUI_CLOCK)   // 3. clock [cite: 18]
      ],
    });

    signAndExecute(
      { transaction: txb as any },
      {
        onSuccess: (result: any) => {
          alert("Quyên góp thành công!");
          setTimeout(() => refetch(), 2000);
        },
        onError: (err: any) => alert("Lỗi: " + err.message)
      }
    );
  };

  return (
    <div className="app-container">
      {showModal && (
        <div className="modal-overlay">
          <div className="modal-content">
            <h2>Tạo chiến dịch mới</h2>
            <input placeholder="Tiêu đề" onChange={e => setFormData({...formData, title: e.target.value})} />
            <input placeholder="Mục tiêu (SUI)" type="number" onChange={e => setFormData({...formData, goal: e.target.value})} />
            <input placeholder="Mô tả ngắn" onChange={e => setFormData({...formData, desc: e.target.value})} />
            <input placeholder="Số ngày kết thúc (VD: 7)" type="number" onChange={e => setFormData({...formData, days: e.target.value})} />
            <div className="modal-btns">
              <button onClick={handleCreateCampaign} className="btn-main">Xác nhận</button>
              <button onClick={() => setShowModal(false)} className="btn-sub">Hủy</button>
            </div>
          </div>
        </div>
      )}

      <nav className="nav-bar">
        <div className="nav-content container">
          <div className="logo">THIỆN<span>NGUYỆN</span></div>
          <div className="nav-right">
            <ConnectButton />
          </div>
        </div>
      </nav>

      <header className="hero-section">
        <div className="container hero-inner">
          <h1>Hệ thống thiện nguyện minh bạch</h1>
          <button className="btn-main" onClick={() => setShowModal(true)}>Tạo chiến dịch ngay</button>
        </div>
      </header>

      <section className="container campaign-section">
        <div className="campaign-grid">
          {campaigns.map(campaign => (
            <div key={campaign.id} className="card">
              <div className="card-content">
                <h3>{campaign.title}</h3>
                <p>{campaign.description}</p>
                <div className="progress-info">
                  <span>{campaign.raised} / {campaign.goal} SUI</span>
                </div>
                <button className="btn-donate-card" onClick={() => handleDonate(campaign.id)}>Ủng hộ 1 SUI</button>
              </div>
            </div>
          ))}
        </div>
      </section>
    </div>
  );
}

export default App;
