// charity-frontend/src/utils/CampaignManager.ts
export interface Campaign {
  id: number;
  title: string;
  description: string;
  image: string;
  raised: number;
  goal: number;
  donors: number;
  category: string;
}

export class CampaignManager {
  private campaigns: Campaign[];
  constructor(campaigns: Campaign[]) { this.campaigns = campaigns; }
  getCampaigns() { return this.campaigns; }
  getProgress(c: Campaign) { return Math.min((c.raised / c.goal) * 100, 100); }
  getTotalRaised() { return this.campaigns.reduce((sum, c) => sum + c.raised, 0); }
}

export const MOCK_DATA: Campaign[] = [
  {
    id: 1,
    title: "Xây dựng lại trường học sau lũ lụt",
    description: "Giúp đỡ các em học sinh vùng cao có nơi học tập tốt hơn sau thiên tai.",
    image: "https://images.unsplash.com/photo-1488521787991-ed7bbaae773c?q=80&w=800",
    raised: 375.5,
    goal: 500,
    donors: 124,
    category: "Giáo dục"
  },
  {
    id: 2,
    title: "Hỗ trợ điều trị cho trẻ em nghèo",
    description: "Quỹ hỗ trợ chi phí phẫu thuật và thuốc men cho các gia đình khó khăn.",
    image: "https://images.unsplash.com/photo-1505155485765-d0f5032b1224?q=80&w=800",
    raised: 150,
    goal: 300,
    donors: 85,
    category: "Y tế"
  }
];
