// charity-frontend/src/components/CampaignCard.tsx
import React from 'react';

interface CampaignCardProps {
  id: number;
  category: string;
  icon: string;
  title: string;
  description: string;
  currentAmount: number;
  goalAmount: number;
  daysLeft: number;
}

const CampaignCard: React.FC<CampaignCardProps> = ({
  category,
  icon,
  title,
  description,
  currentAmount,
  goalAmount,
  daysLeft,
}) => {
  const progress = Math.min((currentAmount / goalAmount) * 100, 100);

  const getCategoryName = (cat: string): string => {
    const categories: { [key: string]: string } = {
      education: 'Giáo dục',
      health: 'Y tế',
      environment: 'Môi trường',
      social: 'Xã hội',
      animals: 'Động vật',
    };
    return categories[cat] || 'Khác';
  };

  const handleDonate = () => {
    // TODO: Open donate modal
    alert(`Ủng hộ cho chiến dịch: ${title}`);
  };

  return (
    <div className="campaign-card" data-category={category}>
      <div className="campaign-image">{icon}</div>
      <div className="campaign-content">
        <span className="campaign-category">{getCategoryName(category)}</span>
        <h3 className="campaign-title">{title}</h3>
        <p className="campaign-description">{description}</p>

        <div className="campaign-progress">
          <div className="progress-bar">
            <div className="progress-fill" style={{ width: `${progress}%` }}></div>
          </div>
          <div className="progress-info">
            <span className="progress-current">{currentAmount} SUI</span>
            <span className="progress-goal">Mục tiêu: {goalAmount} SUI</span>
          </div>
        </div>

        <div className="campaign-footer">
          <span className="campaign-meta">Còn {daysLeft} ngày</span>
          <button className="btn-donate" onClick={handleDonate}>
            Ủng hộ
          </button>
        </div>
      </div>
    </div>
  );
};

export default CampaignCard;
